### ISSUE TITLE
DSA_generate_parameters_ex fails in openssl3

### ISSUE THREAD

**Issue by jeroen** [openssl/openssl#12287]
I know DSA is deprecated, but reading/writing keys still seems to work. However we can no longer generate new DSA keys.Below the code `DSA_generate_parameters_ex` errors with:

```
Error: OpenSSL error: C06DDE0F01000000:error::digital envelope routines:evp_generic_fetch:fetch failed:crypto/evp/evp_fetch.c:307:Default library context, Algorithm (<NULL>), Properties (<null>)
```

I find it confusing that I can still read DSA keys from PEM files, but no longer generate them? Is this intended? 

```c
#include <stdio.h>
#include <openssl/pem.h>

int main( int argc, const char* argv[] ){
  DSA *dsa = DSA_new();
  DSA_generate_parameters_ex(dsa, 2048, NULL, 0, NULL, NULL, NULL);
  DSA_generate_key(dsa);
  exit(0);
}
```

**Comment by levitte** [comment]
Reading a key payload from a file and generating a key payload are two *very* different things.

That being said, I agree that this is a bug.  It seems like some of our code is a bit mixed up about being legacy vs new code.

**Comment by reaperhulk** [comment]
For a bit of additional context: In alpha3 parameter generation works (with or without the legacy provider being loaded). In alpha4 it fails regardless of provider.

**Comment by levitte** [comment]
Good detail to know, thanks

**Comment by mattcaswell** [comment]
I can't reproduce this on master, but I can in alpha4 so it seems this is already fixed. `git bisect` tells me it was fixed by 9beffaf. Can you confirm the issue is fixed for you in master?

**Comment by jeroen** [comment]
Yes it looks like this is fixed. Thanks!
